Add registry for numerical type size and endianness

arteymix opened the pull request.

arteymix committed:
Add initial prototypes for storing byte size and endiannes

arteymix committed:
Add a prototype implementation for the type info table with decimal128

arteymix commented on src/numeric-types.c:
```diff
@@ -93,6 +83,102 @@ numeric_##type##_to_##gtype (numeric_##type num)                     \
+numeric_type_register_static (GType        type,
+                              const gchar *name,
+                              gsize width,
+                              NumericByteOrder byte_order)
```
The alternative is to expose the `NumericTypeInfo` structure so we can use a single argument for the metadata.

arteymix commented on src/numeric-types.c:
```diff
@@ -112,11 +198,18 @@ DEFINE_NUMERIC_WITH_BYTESWAP (double_be, double, gint32,  GINT64,  BE)
+                                  sizeof (numeric_decimal128),
+                                  NUMERIC_BYTE_ORDER_UNKNOWN);
+
+#if HAVE_LIBDFP
```
Registration should be performed lazily in `numeric_{type}_get_type`, but that would require redefining GLib macros for boxed types registration.

arteymix commented on src/numeric-types.c:
```diff
@@ -93,6 +83,102 @@ numeric_##type##_to_##gtype (numeric_##type num)                     \
+    GType             type;
+    const gchar      *name;
+    gsize             width;
+    NumericByteOrder  byte_order;
```
@rainwoodman If you can think of any other field that could be of use in Vast for example.

If we know the width and byte ordering, we could very likely generate the code for byte swapping.

rainwoodman commented:
I suggest making name a fixed size array.

These are scalar types, so I dont think we want to add vector size or
composite types here.

There is a standard byte order swapping algorithm, e.g.

https://github.com/rainwoodman/bigfile/blob/41623783e61a6971bb1ab4a09cc3d797b6d90283/src/bigfile.c#L1297

I think I took that from numpy.


On Thu, Feb 13, 2020, 1:34 PM Guillaume Poirier-Morency <
notifications@github.com> wrote:

> *@arteymix* commented on this pull request.
> ------------------------------
>
> In src/numeric-types.c
> <https://github.com/arteymix/numeric-glib/pull/17#discussion_r379131082>:
>
> > @@ -93,6 +83,102 @@ numeric_##type##_to_##gtype (numeric_##type num)                     \
>      return data.v_##gtype;                                           \
>  }
>
> +typedef struct _NumericTypeInfo
> +{
> +    GType             type;
> +    const gchar      *name;
> +    gsize             width;
> +    NumericByteOrder  byte_order;
>
> @rainwoodman <https://github.com/rainwoodman> If you can think of any
> other field that could be of use in Vast for example.
>
> If we know the width and byte ordering, we could very likely generate the
> code for byte swapping.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/arteymix/numeric-glib/pull/17?email_source=notifications&email_token=AABBWTEHZJPVA7UWKI5DAYLRCW4F5A5CNFSM4KU2OG6KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVPU3EI#pullrequestreview-358567313>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABBWTE6NE6JFRXL6LZOA2TRCW4F5ANCNFSM4KU2OG6A>
> .
>

arteymix committed:
Register all NumericTypeInfo structs and reuse GLib byte order macros

Use G_DEFINE_BOXED_TYPE_WITH_CODE to register numeric typeinfo
counterpart lazily in the *_get_type routine.

Reuse GLib byte order macros since our definitions are at best
redundant.

arteymix commented on src/numeric-types.c:
```diff
@@ -18,13 +18,15 @@
+                               numeric_##type,                   \
+                               numeric_##type##_copy,            \
+                               numeric_##type##_free,            \
+                               numeric_type_register_static (g_define_type_id, G_STRINGIFY(type), sizeof (numeric_##type), G_BYTE_ORDER)) \
```
We have to add `byte_order` to the `DEFINE_NUMERIC` macro because not all definitions use host order.

arteymix commented on src/numeric-types.c:
```diff
@@ -93,6 +84,117 @@ numeric_##type##_to_##gtype (numeric_##type num)                     \
+numeric_type_get_name (GType numeric_type)
+{
+    const NumericTypeInfo *type_info;
+    type_info = numeric_get_type_info (numeric_type);
```
Add `g_return_val_if_fail (type_info != NULL, NULL)` here and below.

arteymix commented on src/numeric-types.c:
```diff
@@ -112,11 +214,25 @@ DEFINE_NUMERIC_WITH_BYTESWAP (double_be, double, gint32,  GINT64,  BE)
+    numeric_type_register_static (G_TYPE_INT64,  "int64",  sizeof (gint64),  G_BYTE_ORDER);
+    numeric_type_register_static (G_TYPE_UINT64, "uint64", sizeof (guint64), G_BYTE_ORDER);
+    numeric_type_register_static (G_TYPE_FLOAT,  "float",  sizeof (gfloat),  G_BYTE_ORDER);
+    numeric_type_register_static (G_TYPE_DOUBLE, "double", sizeof (gdouble), G_BYTE_ORDER);
```
The issue here is that we cannot redefine these types. We can either have a numeric boxed counterpart or look into switching from boxed types to fundamental types for our definitions.

The quickfix would be to define these "fundamental" types lazily in `numeric_get_type_info`.

rainwoodman replied:
2 quick points: Would life be made easier if not using macros but hard code and/or generate the definitions with a code generator?

I remember the get_type() functions always generate the type the first time. Is that what you gain by moving to a fundamental type? If adding a fundamental type. I would suggest considering GNumericScalar as an umbrella type for all of the types.

arteymix replied:
It's actually doing just that! `G_DEFINE_BOXED_TYPE_WITH_CODE` inserts arbitrary code inside the `_get_type` definition allowing us to perform any kind of initialization and registration of transformations.

The issue I'm facing with GLib fundamental types is that we cannot redefine that piece of code and thus registering the `NumericTypeInfo` counterpart we will need for introspection and byteswapping.

arteymix commented on src/numeric-types.c:
```diff
@@ -93,6 +84,117 @@ numeric_##type##_to_##gtype (numeric_##type num)                     \
     return data.v_##gtype;                                           \
 }
 
+GArray *TYPE_TABLE = NULL;
```
Make this declaration static.

arteymix committed:
Remove dynamic initialization of GLib numerical types and simplify API

Let numeric_type_register_static take the NumericTypeInfo struct as
argument.

Move registration of numeric counterparts of GLib fundamental types into
numeric_get_type_info().

Mark the TYPE_TABLE symbol as static.

Remove _get_name, _get_type and _get_byte_order in favour of using the
now public NumericTypeInfo struct members directly.

arteymix commented on src/numeric-types.c:
```diff
@@ -93,6 +84,114 @@ numeric_##type##_to_##gtype (numeric_##type num)                     \
+    numeric_type_register_static_simple (G_TYPE_UINT64, "uint64", sizeof (guint64), G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_FLOAT,  "float",  sizeof (gfloat),  G_BYTE_ORDER);
+    numeric_type_register_static_simple (G_TYPE_DOUBLE, "double", sizeof (gdouble), G_BYTE_ORDER);
+}
```
It would be great if we could enumerate these types at runtime so that the we could be forward compatibly if new types are defined in GLib.

arteymix commented on src/numeric-glib-1.0.vapi:
```diff
@@ -19,6 +19,19 @@
+	}
+
+	public unowned Numeric.TypeInfo get_type_info (GLib.Type type);
+	public unowned Numeric.TypeInfo get_type_info_from_name (string name);
```
These functions should be defined within the `TypeInfo` class as static.

arteymix committed:
Fix byte order information for _le and _be flavours

arteymix commented on src/numeric-types.c:
```diff
@@ -93,30 +84,140 @@ numeric_##type##_to_##gtype (numeric_##type num)                     \
+DEFINE_NUMERIC_WITH_BYTESWAP (float_le,  G_LITTLE_ENDIAN, float,  gint32,  GINT32,  LE)
+DEFINE_NUMERIC_WITH_BYTESWAP (float_be,  G_BIG_ENDIAN,    float,  gint32,  GINT32,  BE)
+DEFINE_NUMERIC_WITH_BYTESWAP (double_le, G_LITTLE_ENDIAN, double, gint32,  GINT64,  LE)
+DEFINE_NUMERIC_WITH_BYTESWAP (double_be, G_BIG_ENDIAN,    double, gint32,  GINT64,  BE)
```
I think we're almost good now. I will incorporate your piece of code (if license permit?) to perform generic byte swapping.

rainwoodman replied:
No problem. There is a minimal version of  numpy 'dtype' in bigfile, included some macros for casting and striding. Feel free to recycle any pieces there.

arteymix committed:
Add init/clear for NumericTypeInfo

It will be useful later on when we integrate dynamic types.
